allennlp/semparse/worlds/atis_world.py

```diff
@@ -0,0 +1,133 @@
+        and numbers that were valid earlier in the interaction are also valid. We then add new valid strings
+        and numbers from the current utterance.
+        """
+        valid_actions = deepcopy(self.sql_table_context.valid_actions)
+        for string in self.get_strings_from_utterance():
+            if string not in valid_actions['string']:
+                valid_actions['string'].append(format_action('string', string))
+
+        numbers = ['0', '1']
+        for utterance in self.utterances:
+            numbers.extend(get_numbers_from_utterance(utterance))
+            for number in numbers:
+                if number not in valid_actions['number']:
+                    valid_actions['number'].append(format_action('number', number))
+
+        return valid_actions
+
+    def get_grammar_str(self) -> str:
+        """
+        Generate a string that can be used to instantiate a ``Grammar`` object. The string is a sequence of
+        rules that define the grammar.
+        """
+        grammar_str_with_context = self.sql_table_context.grammar_str
+        numbers = [number.split(" -> ")[1].lstrip('["').rstrip('"]') for \
+                   number in sorted(self.valid_actions['number'], reverse=True)]
+        strings = [string .split(" -> ")[1].lstrip('["').rstrip('"]') for \
+                   string in sorted(self.valid_actions['string'], reverse=True)]
+
+        grammar_str_with_context += generate_one_of_string("number", numbers)
+        grammar_str_with_context += generate_one_of_string("string", strings)
+        return grammar_str_with_context
+
+
+    def get_strings_from_utterance(self) -> List[str]:
+        """
+        Based on the current utterance, return a list of valid strings that should be added.
+        """
+        strings: List[str] = []
+        trigger_lists = [CITIES, AIRPORT_CODES,
+                         STATES, STATE_CODES,
+                         FARE_BASIS_CODE, CLASS,
+                         AIRLINE_CODE_LIST, DAY_OF_WEEK,
+                         CITY_CODE_LIST, MEALS,
+                         RESTRICT_CODES]
+        trigger_dicts = [CITY_AIRPORT_CODES,
+                         AIRLINE_CODES,
+                         CITY_CODES,
+                         GROUND_SERVICE,
+                         YES_NO,
+                         MISC_STR]
+
+        trigger_dict = get_trigger_dict(trigger_lists, trigger_dicts)
+
+        for tokenized_utterance in self.tokenized_utterances:
+            if tokenized_utterance:
+                for first_token, second_token in zip(tokenized_utterance, tokenized_utterance[1:]):
+                    strings.extend(trigger_dict.get(first_token.text.lower(), []))
+                    bigram = f"{first_token.text} {second_token.text}"
+                    strings.extend(trigger_dict.get(bigram.lower(), []))
+                strings.extend(trigger_dict.get(tokenized_utterance[-1].text.lower(), []))
+
+        strings.extend(DAY_OF_WEEK)
```

Comment:
You're always allowing all days of the week?  Why?  Because you don't know a priori what day it is?

Reply:
There are many examples where the utterance mentions a date and the corresponding SQL query has a day of week. For example `show me flights from denver to philadelphia on a 1991 september second"` has the string `MONDAY` in its SQL query. I'll add a method for getting the date, and then adding the day of week string based on that.